### ISSUE TITLE
Create a clearer division between the API and the Runtime

### ISSUE THREAD

**Issue by DisabledMallis** [FrederoxDev/Amethyst#38]
Right now it is not exactly clear what the API and Runtime parts of Amethyst are actually for. What I mean is that the API contains classes like HookManager and InputManager, while the Runtime appears to be whats responsible for making them work...

I think this is fine, having common data, structures, and functionality in the Amethyst API, as well as initializing hooks and executing input callbacks in the Runtime - but I think it should be a cleaner setup to make it clear what exactly everything is.

Most modding platforms such as forge offer an "Event Bus", and I think amethyst should have something similar. From my understanding, Amethyst is supposed to be like Fabric but for Bedrock edition, so having mods initialize their own hooks is reasonable, but I think having a central "Event Bus" system should also be provided to help mod compatibility.

On top of all of this, Hook Manager, Config, Runtime Input Manager are all members of the AmethystRuntime class - why?

Mod functions are a strange concept as well, this is functionality that can be replaced with the event bus idea (with the exception of mod initialization and shutdown, for example).

The solution to this I believe is to have a central AmethystAPI class that contains a map of a type id hash, and a corresponding instance. This would be for quick access to the various API feature singletons, such as Hook Manager. The API is never responsible for creating hooks, but it, in my opinion, should be responsible for managing them.

On top of this, I think there should be a central system for getting the addresses of various symbols and class member offsets. The runtime would be responsible for actually finding these addresses and offsets with the use of signatures. The API can then be written to use these addresses and offsets provided by the Runtime in a central way, allowing for the API itself to be almost versionless in a way. The Runtime would be the only thing that needs to be 'updated', with the exception if Mojang removes a class member or function.

In conclusion, the Runtime itself would be solely responsible for finding functions and offsets. The API would be responsible for managing the results from the Runtime, as well as providing a central means to manage hooks, events, and provide classes, structs and functions for Mods.

**Comment by FrederoxDev** [comment]
The current thing holding us back from this right now is the launcher: https://github.com/FrederoxDev/Amethyst/issues/32 as in its current state, some hooks are time dependent so it didn't make complete sense to have two separate things. A better launcher that injects earlier is currently being worked on

**Comment by FrederoxDev** [comment]
It's certainly not everything, but this commit moves away from using ModFunctions in favor for creating callbacks https://github.com/FrederoxDev/Amethyst/commit/4b560ebfdacfd0309365bb5c4bfc3e4e309570a2

**Comment by FrederoxDev** [comment]
https://github.com/FrederoxDev/Amethyst/commit/734a64e993910493f5b3bff9021502402623d764
